R/install.R

```diff
@@ -1,49 +1,68 @@
+#'   environment.
+#' @param new_env If `TRUE`, any existing Python virtual environment and/or
+#'   Conda environment specified by `envname` is deleted first.
+#' @param method The installation method to use. If creating a new environment,
+#'   `"auto"` (the default) is equivalent to `"virtualenv"`. Otherwise `"auto"`
+#'   infers the installation method based on the type of Python environment
+#'   specified by `envname`.
+#' @param ... Passed on to [`reticulate::py_install()`]
 #' @export
-install_pyspark <- function(virtualenv_name = "r-sparklyr",
-                            python_version = NULL,
-                            force = FALSE,
-                            ignored_installed = TRUE,
-                            method = c("auto", "virtualenv", "conda")
-                            ) {
-
-  version <- ">=3.9"
-
-  pkgs <- c(
-    "pyspark", "pandas", "PyArrow", "grpcio", "google-api-python-client",
-    "grpcio_status", "databricks-connect", "delta-spark"
+install_pyspark <- function(envname = "r-sparklyr",
+                            ...,
+                            python_version = ">=3.9",
+                            new_env = identical(envname, "r-sparklyr"),
+                            method = c("auto", "virtualenv", "conda")) {
+
+  packages <- c(
+    "pyspark",
+    "pandas",
+    "PyArrow",
+    "grpcio",
+    "google-api-python-client",
+    "grpcio_status",
+    "databricks-connect",
+    "delta-spark"
   )
 
-  opts <- "--index-url https://packagemanager.posit.co/pypi/2023-06-01/simple"
+  pip_options <- "--index-url https://packagemanager.posit.co/pypi/2023-06-01/simple"
+  # in cause user supplied pip_options in ...
+  pip_options <- c(pip_options, list(...)$pip_options)
+
+  method <- match.arg(method)
 
-  if(is.null(virtualenv_starter(version))) {
-    cli_abort(
-      paste(
+  if(new_env) {
+    if(method %in% c("auto", "virtualenv"))
+      tryCatch(virtualenv_remove(envname, confirm = FALSE), error = identity)
+    if(method %in% c("auto", "conda"))
+      tryCatch(conda_remove(envname, conda = list(...)$conda %||% "auto"),
+                            error = identity)
+  }
+
+  if(new_env && method != "conda" &&
+     is.null(virtualenv_starter(python_version)))
+    cli_abort(paste(
         "Python version 3.9 or higher is required by some libraries.",
         "Use: {.run reticulate::install_python(version = '3.9:latest')}",
         "to install."
-        )
-      )
-  }
+    ))
+
+
+  # conda_install() doesn't accept a version constraint for python_version
+  if(method == "conda" && python_version == ">=3.9")
+    python_version <- "3.9"
 
   py_install(
-    packages = pkgs,
-    envname = virtualenv_name,
-    ignore_installed = ignored_installed,
-    pip_options = opts,
+    packages = packages,
+    envname = envname,
     method = method,
-    version = version,
-    force = force
-    )
+    python_version = python_version,
+    pip = TRUE,
```

Comment:
This interesting Derrick had an issue with `pip` in his installation, it seems that it was missing, I wonder if this switch addresses that